ext/hook/uhook.c

```diff
@@ -0,0 +1,331 @@
+    ZVAL_OBJ(&hook_data_zv, hook_data);
+
+    bool has_this = getThis() != NULL;
+    zval rv;
+    zai_symbol_call(has_this ? ZAI_SYMBOL_SCOPE_OBJECT : ZAI_SYMBOL_SCOPE_GLOBAL, has_this ? &EX(This) : NULL, ZAI_SYMBOL_FUNCTION_CLOSURE, &closure_zv, &rv, 1, &hook_data_zv);
+    zval_ptr_dtor(&rv);
+}
+
+static bool dd_uhook_begin(zend_ulong invocation, zend_execute_data *execute_data, void *auxiliary, void *dynamic) {
+    (void) invocation;
+
+    dd_uhook_def *def = auxiliary;
+    dd_uhook_dynamic *dyn = dynamic;
+
+    dyn->hook_data = zend_objects_new(ddtrace_hook_data_ce);
+    object_properties_init(dyn->hook_data, ddtrace_hook_data_ce);
+
+    ZVAL_LONG(ddtrace_hookdata_property_id(dyn->hook_data), def->id);
+    ZVAL_ARR(ddtrace_hookdata_property_args(dyn->hook_data), dd_uhook_collect_args(execute_data));
+
+    if (def->begin && !def->running) {
+        def->running = true;
+        dd_uhook_call_hook(execute_data, def->begin, dyn->hook_data);
+        def->running = false;
+    }
+
+    return true;
+}
+
+static void dd_uhook_end(zend_ulong invocation, zend_execute_data *execute_data, zval *retval, void *auxiliary, void *dynamic) {
+    (void) invocation;
+
+    dd_uhook_def *def = auxiliary;
+    dd_uhook_dynamic *dyn = dynamic;
+
+    if (def->end && !def->running) {
+        zval tmp;
+
+        /* If the profiler doesn't handle a potential pending interrupt before
+         * the observer's end function, then the callback will be at the top of
+         * the stack even though it's not responsible.
+         * This is why the profiler's interrupt function is called here, to
+         * give the profiler an opportunity to take a sample before calling the
+         * tracing funcation.
+         */
+        if (profiling_interrupt_function) {
+            profiling_interrupt_function(execute_data);
+        }
+
+        zval *returned = ddtrace_hookdata_property_returned(dyn->hook_data);
+        ZVAL_COPY_VALUE(&tmp, returned);
+        ZVAL_COPY(returned, retval);
+        zval_ptr_dtor(&tmp);
+
+        zval *exception = ddtrace_hookdata_property_exception(dyn->hook_data);
+        ZVAL_COPY_VALUE(&tmp, exception);
+        if (EG(exception)) {
+            ZVAL_OBJ_COPY(exception, EG(exception));
+        } else {
+            ZVAL_NULL(exception);
+        }
+        zval_ptr_dtor(&tmp);
+
+        def->running = true;
+        dd_uhook_call_hook(execute_data, def->end, dyn->hook_data);
+        def->running = false;
+    }
+
+    OBJ_RELEASE(dyn->hook_data);
+}
+
+static void dd_uhook_dtor(void *data) {
+    dd_uhook_def *def = data;
+    if (def->begin) {
+        OBJ_RELEASE(def->begin);
+    }
+    if (def->end) {
+        OBJ_RELEASE(def->end);
+    }
+    if (def->function) {
+        zend_string_release(def->function);
+        if (def->scope) {
+            zend_string_release(def->scope);
+        }
+    }
+    zend_hash_index_del(&dd_active_hooks, (zend_ulong)def->id);
+    efree(def);
+}
+
+// clang-format off
```

Comment:
Why?

Reply:
(I have no idea, it's not mine, I believe it persisted through some rewriting of the code)